subprojects/gradle/buildchecks/src/main/kotlin/com/avito/android/plugin/build_param_check/BuildParamCheckPlugin.kt

```diff
@@ -16,70 +17,118 @@ import org.gradle.api.Plugin
+            project.gradle.startParameter.taskNames + "checkBuildEnvironment"
+        )
+
+        if (checks.hasInstance<Check.AndroidSdk>()) {
+            val check = checks.getInstance<Check.AndroidSdk>()
+            val task = project.tasks.register<CheckAndroidSdkVersionTask>("checkAndroidSdkVersion") {
                 group = "verification"
                 description =
                     "Checks sdk version in docker against local one to prevent build cache misses"
 
+                compileSdkVersion.set(check.compileSdkVersion)
+                revision.set(check.revision)
                 // don't run task if it is already compared hashes and it's ok
                 // task will be executed next time if either local jar or docker jar(e.g. inputs) changed
                 outputs.upToDateWhen { outputs.files.singleFile.exists() }
-                onlyIf { enabled }
             }
-        val checkGradleDaemonTask = project.tasks.register<CheckGradleDaemonTask>("checkGradleDaemon") {
-            group = "verification"
-            description = "Check gradle daemon problems"
+            checkBuildEnvironment {
+                dependsOn(task)
+            }
+        }
+        if (checks.hasInstance<Check.GradleDaemon>()) {
+            val task = project.tasks.register<CheckGradleDaemonTask>("checkGradleDaemon") {
+                group = "verification"
+                description = "Check gradle daemon problems"
+            }
+            checkBuildEnvironment {
+                dependsOn(task)
+            }
         }
-        val dynamicDependenciesTask = project.tasks.register<DynamicDependenciesTask>("checkDynamicDependencies") {
-            group = "verification"
-            description = "Detects dynamic dependencies"
+        if (checks.hasInstance<Check.DynamicDependencies>()) {
+            val task = project.tasks.register<DynamicDependenciesTask>("checkDynamicDependencies") {
+                group = "verification"
+                description = "Detects dynamic dependencies"
+            }
+            checkBuildEnvironment {
+                dependsOn(task)
+            }
         }
-        val checkBuildEnvironment = project.tasks.register("checkBuildEnvironment") {
-            it.group = "verification"
-            it.description = "Check typical build problems"
-            it.dependsOn(checkAndroidSdk, checkGradleDaemonTask, dynamicDependenciesTask)
+        if (checks.hasInstance<Check.UniqueRClasses>()) {
+            check(project.pluginManager.hasPlugin("com.avito.android.impact")) {
+                "build check for unique R classes requires 'com.avito.android.impact' plugin"
+            }
+            val task = project.tasks.register<UniqueRClassesTask>("checkUniqueAndroidPackages") {
+                group = "verification"
+                description = "Verify unique R classes"
+            }
+            checkBuildEnvironment {
+                dependsOn(task)
```

Comment:
Did you measure of time it spends? Not only a new one, the whole configuration phase with and without checks

Reply:
We only register a new task here.

The plugin does some checks at the configuration phase. Gradle build scan shows up to 0.5 for this plugin. Definitely, we can improve it, but not in this PR.

Reply:
`UniqueRClassesTask` itself lasts about 5s

Reply:
Do checks run in execution phase?

Reply:
Yes, the are Gradle tasks

```
checkBuildEnvironment { <--- task provider, "{" is a function for ".configure"
                dependsOn(task) <--- task dependency
```

Reply:
Could we make them cacheble? Or is it overwork?

Reply:
It would be good. It depends.

For instance, checking localhost in macOs don't have inputs in the project. It's hard to say when to cache.

Another example: unique R classes. The input of the task is the whole project structure and manifests.

The first step is to migrate to the worker API. I think It will be enough.